src/Feldspar/Compiler/Imperative/Frontend.hs

```diff
@@ -56,56 +56,63 @@ copyProg Nothing _ = Empty
+  | not (isArray (typeof arg1) || isNativeArray (typeof arg1))
+  = Assign arg1 arg2
+    where deepCopyField fld = deepCopy (StructField arg1 fld) [StructField arg2 fld]
 
-deepCopy _ (ValueParameter arg1 : ins'@(ValueParameter in1:ins))
-  | isArray (typeof arg1)
-  = [ initArray (Just arg1) expDstLen, copyFirstSegment ] ++
-      flattenCopy (ValueParameter arg1) ins argnLens arg1len
-    where expDstLen = foldr ePlus (litI32 0) aLens
-          copyFirstSegment = if arg1 == in1
-                                then Empty
-                                else call "copyArray" [ ValueParameter arg1
-                                                      , ValueParameter in1]
-          aLens@(arg1len:argnLens) = map (\(ValueParameter src) -> arrayLength src) ins'
+deepCopy arg1 args
+  | isArray (typeof arg1) || isNativeArray (typeof arg1)
+  = Assign arg1 $ fun (typeof arg1) "copy" (arg1 : args)
 
 deepCopy _ _ = error "Multiple non-array arguments to copy"
 
-flattenCopy :: ActualParameter () -> [ActualParameter ()] -> [Expression ()] ->
+flattenCopy :: Expression () -> [Expression ()] -> [Expression ()] ->
                Expression () -> [Program ()]
 flattenCopy _ [] [] _ = []
-flattenCopy dst (t:ts) (l:ls) cLen = call "copyArrayPos" [dst, ValueParameter cLen, t]
+flattenCopy dst (t:ts) (l:ls) cLen = call "copyArrayPos" [ValueParameter dst, ValueParameter cLen, ValueParameter t]
                                    : flattenCopy dst ts ls (ePlus cLen l)
 
 ePlus :: Expression () -> Expression () -> Expression ()
 ePlus (ConstExpr (IntConst 0 _)) e = e
 ePlus e (ConstExpr (IntConst 0 _)) = e
 ePlus e1 e2                        = binop (1 :# (NumType Signed S32)) "+" e1 e2
 
+-- | Lower array copy
+lowerCopy :: Options -> Type -> Expression () -> [Expression ()] -> [Program ()]
+lowerCopy opts ArrayType{} arg1 ins'@(in1:ins)
+  | [ConstExpr ArrayConst{..}] <- ins'
+  = initArray (Just arg1) (litI32 $ toInteger $ length arrayValues)
+    : zipWith (\i c -> Assign (ArrayElem arg1 [litI32 i]) (ConstExpr c)) [0..] arrayValues
+
+  | otherwise
+  = [ initArray (Just arg1) expDstLen, copyFirstSegment ] ++
+      flattenCopy arg1 ins argnLens arg1len
+    where expDstLen = foldr ePlus (litI32 0) aLens
+          copyFirstSegment = if arg1 == in1
+                                then Empty
+                                else call "copyArray" [ ValueParameter arg1
+                                                      , ValueParameter in1]
+          aLens@(arg1len:argnLens) = map (\src -> arrayLength src) ins'
+lowerCopy opts NativeArray{} arg1 [arg2]
+  | l@(ConstExpr (IntConst n _)) <- arrayLength arg2
+  = if n < safetyLimit opts
+      then initArray (Just arg1) l:map (\i -> Assign (ArrayElem arg1 [litI32 i]) (ArrayElem arg2 [litI32 i])) [0..(n-1)]
+      else error $ unlines ["Internal compiler error: array size (" ++ show n ++ ") too large for deepcopy", show arg1, show arg2]
```

Comment:
Use "lowerCopy" to refer to the function failing in the error message.

Reply:
Yup.